xmake/rules/c++/modules/xmake.lua

```diff
@@ -72,116 +38,31 @@ rule("c++.build.modules.builder")
-            sourcebatch.objectfiles = {}
-        end
+        import("builder")(target, jobgraph, sourcebatch, opt)
     end, {jobgraph = true, batch = true})
 
     -- serial compilation only, usually used to support project generator
     before_buildcmd_files(function(target, batchcmds, sourcebatch, opt)
-        if target:data("cxx.has_modules") then
-            import("support")
-            import("scanner")
-            import("builder")
-
-            -- get module dependencies
-            local modules = scanner.get_module_dependencies(target, sourcebatch)
-            if not target:is_moduleonly() then
-                -- avoid building non referenced modules
-                local build_objectfiles, link_objectfiles = scanner.sort_modules_by_dependencies(target, sourcebatch.objectfiles, modules)
-                sourcebatch.objectfiles = build_objectfiles
-
-                -- build headerunits and modules
-                builder.build_modules_and_headerunits(target, batchcmds, sourcebatch, modules, opt)
-                sourcebatch.objectfiles = link_objectfiles
-            else
-                -- avoid duplicate linking of object files of non-module programs
-                sourcebatch.objectfiles = {}
-            end
-
-            support.localcache():set2(target:fullname(), "c++.modules", modules)
-            support.localcache():save()
-        else
-            sourcebatch.sourcefiles = {}
-            sourcebatch.objectfiles = {}
-            sourcebatch.dependfiles = {}
-        end
+        import("builder")(target, batchcmds, sourcebatch, opt)
     end)
 
     after_clean(function (target)
-        import("core.base.option")
-        import("support")
-        import("private.action.clean.remove_files")
-
-        -- we cannot use target:data("cxx.has_modules"),
-        -- because on_config will be not called when cleaning targets
-        if support.contains_modules(target) then
-            remove_files(support.modules_cachedir(target))
-            if option.get("all") then
-                remove_files(support.stlmodules_cachedir(target))
-                support.localcache():clear()
-                support.localcache():save()
-            end
-        end
+        import("builder").clean(target)
     end)
 
 -- install modules
 rule("c++.build.modules.install")
     set_extensions(".mpp", ".mxx", ".cppm", ".ixx")
 
     before_install(function (target)
-        import("support")
-        import("builder")
-
-        -- we cannot use target:data("cxx.has_modules"),
-        -- because on_config will be not called when installing targets
-        if support.contains_modules(target) then
-            local modules = support.localcache():get2(target:fullname(), "c++.modules")
-            builder.generate_metadata(target, modules)
-
-            support.add_installfiles_for_modules(target)
-        end
+        import("builder").install(target)
```

Comment:
here

Reply:
this one i don't call main but install function

Reply:
oh nice ok